rustls/examples/internal/test_ca.rs

```diff
@@ -0,0 +1,321 @@
+
+            // If we're not a trust anchor, generate a CRL for the certificate we just issued.
+            if !matches!(role, Role::TrustAnchor) {
+                // The CRL will be signed by the issuer of the certificate being revoked. For
+                // intermediates this will be the trust anchor, and for client/EE certs this will
+                // be the intermediate.
+                let issuer = match role {
+                    Role::Intermediate => certified_keys
+                        .get(&(Role::TrustAnchor, alg.inner))
+                        .unwrap(),
+                    Role::EndEntity | Role::Client => certified_keys
+                        .get(&(Role::Intermediate, alg.inner))
+                        .unwrap(),
+                    _ => panic!("unexpected role for CRL generation: {role:?}"),
+                };
+                let crl = crl_for_serial(
+                    cert.params()
+                        .serial_number
+                        .as_ref()
+                        .unwrap(),
+                )
+                .signed_by(&issuer.cert, &issuer.key_pair)?;
+                let mut crl_file = File::create(
+                    alg.output_directory()
+                        .join(format!("{}.revoked.crl.pem", role.label())),
+                )?;
+                crl_file.write_all(crl.pem().unwrap().as_bytes())?;
+            }
+
+            // When we're issuing end entity or client certs we have a bit of extra work to do
+            // now that we have full chains in hand.
+            if matches!(role, Role::EndEntity | Role::Client) {
+                let root = &certified_keys
+                    .get(&(Role::TrustAnchor, alg.inner))
+                    .unwrap()
+                    .cert;
+                let intermediate = &certified_keys
+                    .get(&(Role::Intermediate, alg.inner))
+                    .unwrap()
+                    .cert;
+
+                // Write the PEM chain and full chain files for the end entity and client certs.
+                // Chain files include the intermediate and root certs, while full chain files include
+                // the end entity or client cert as well.
+                for f in [
+                    ("chain", &[intermediate, root][..]),
+                    ("fullchain", &[&cert, intermediate, root][..]),
+                ] {
+                    let mut chain_file = File::create(alg.output_directory().join(format!(
+                        "{}.{}",
+                        role.label(),
+                        f.0
+                    )))?;
+                    for cert in f.1 {
+                        chain_file.write_all(cert.pem().as_bytes())?;
+                    }
+                }
+            }
+
+            certified_keys.insert((role, alg.inner), CertifiedKey { cert, key_pair });
+        }
+    }
+
+    Ok(())
+}
+
+fn crl_for_serial(serial_number: &SerialNumber) -> CertificateRevocationListParams {
```

Comment:
Nit: maybe consider taking `SerialNumber` directly here, moving the `clone()` to the outside? The caller can just use `cloned()` instead of `as_ref()` so that's a small difference.

Reply:
Good idea, fixed.